### DISCUSSION TITLE
missing features in sdf generation

### DISCUSSION THREAD

**Discussion by SirNova01** [libigl/libigl#1842] [Q&A]
I tried generating the sdf for a .obj mesh and also ran marching cubes algorithm on it using libigl but the output was really rough and unexpected, i seem to be getting something wrong... please help!
here's my code

#include <igl/marching_cubes.h>
#include <igl/signed_distance.h>
#include <igl/read_triangle_mesh.h>
#include <igl/voxel_grid.h>
#include <igl/opengl/glfw/Viewer.h>
#include <Eigen/Core>
#include <iostream>
#include <igl/parula.h>

#include "tutorial_shared_path.h"

int main(int argc, char * argv[])
{
  using namespace Eigen;
  using namespace std;
  using namespace igl;
  MatrixXi F;
  MatrixXd V;
  Eigen::MatrixXd C;
  // Read in inputs as double precision floating point meshes
  read_triangle_mesh("armadillo.obj",V,F);
  // igl::readOBJ("armadillo.obj", V, F);
  cout<<"Creating grid..."<<endl;
  // number of vertices on the largest side
  const int s = 64;
  // create grid
  MatrixXd GV;
  Eigen::RowVector3i res;
  igl::voxel_grid(V,0,s,1,GV,res);
 
  // compute values
  cout<<"Computing distances..."<<endl;
  VectorXd S,B;
  {
    VectorXi I;
    MatrixXd C,N;
    signed_distance(GV,V,F,SIGNED_DISTANCE_TYPE_PSEUDONORMAL,S,I,C,N);
    // signed_distance(GV,V,F,SIGNED_DISTANCE_TYPE_FAST_WINDING_NUMBER,S,I,C,N);
    // Convert distances to binary inside-outside data --> aliasing artifacts
    B = S;
    for_each(B.data(),B.data()+B.size(),[](double& b){b=(b>0?1:(b<0?-1:0));});
  }
  cout<<"Marching cubes..."<<endl;
  MatrixXd SV,BV,AV;
  MatrixXi SF,BF,AF;
  igl::marching_cubes(S,GV,res(0),res(1),res(2),0,SV,SF);
  igl::marching_cubes(B,GV,res(0),res(1),res(2),0,BV,BF);

  cout<<R"(Usage:
  '1'  Show original mesh.
  '2'  Show marching cubes contour of signed distance.
  '3'  Show marching cubes contour of indicator function.
  )";
  igl::opengl::glfw::Viewer viewer;
  viewer.data().set_mesh(SV,SF);

  ofstream write2file("output.txt");
    if(write2file.is_open())
    {
      write2file << S;
      write2file.close();
      // std::this_thread::sleep_for(std::chrono::milliseconds(2000));
      cout<<"Output is SAVE at "<<"output.txt"<<endl;
    }

  viewer.callback_key_down =
    [&](igl::opengl::glfw::Viewer & viewer, unsigned char key, int mod)->bool
    {
      switch(key)
      {
        default:
          return false;
        case '1':
          viewer.data().clear();
          viewer.data().set_mesh(V,F);
          break;
        case '2':
          viewer.data().clear();
          viewer.data().set_mesh(SV,SF);
          break;
        case '3':
          viewer.data().clear();
          viewer.data().set_mesh(BV,BF);
          break;
        case '4':
          viewer.data().clear();
          viewer.data().set_mesh(SV,SF);
          break;
      }
      viewer.data().set_face_based(true);
      return true;
    };

     // Use the (normalized) vertex positions as colors
  // C = 
  //   (SV.rowwise()            - SV.colwise().minCoeff()).array().rowwise()/
  //   (SV.colwise().maxCoeff() - SV.colwise().minCoeff()).array();

  // Add per-vertex colors
  // viewer.data().set_colors(C);

  viewer.launch();
}

also: here is a link to the mesh file i used. it's a cabinet triangular mesh from shapenet
https://drive.google.com/file/d/17PwifnzyzIlTI1vKN8dm_FYb9fvdUyoE/view?usp=sharing

these were the results i got
<h4>Generated mesh after marching cubes</h4>
<img width="1280" alt="Screen Shot 2021-07-06 at 9 03 30 AM" src="https://user-images.githubusercontent.com/49219813/124564986-7f967100-de39-11eb-9795-dc390eea2aa2.png">
<h4>Original Mesh </h4>
<img width="1280" alt="Screen Shot 2021-07-06 at 9 03 35 AM" src="https://user-images.githubusercontent.com/49219813/124564876-67265680-de39-11eb-89ab-59cc429406a5.png">

**Comment by jdumas** [comment]
Have you tried with the winding numbers /  a higher resolution?

**Reply by SirNova01** [reply]
Hi thanks for your response, 
i do not really understand what you meant by "trying out the winding numbers"
could you kindly explain please

**Reply by jdumas** [reply]
These two lines in your code sample:
```cpp
signed_distance(GV,V,F,SIGNED_DISTANCE_TYPE_PSEUDONORMAL,S,I,C,N);
// signed_distance(GV,V,F,SIGNED_DISTANCE_TYPE_FAST_WINDING_NUMBER,S,I,C,N);
```

**Comment by ckingkan** [comment]
A quick solution would be to increase your grid resolution 

```
const int s = 64; -> count int s=128;
// create grid
MatrixXd GV;
Eigen::RowVector3i res;
igl::voxel_grid(V,0,s,1,GV,res);
```

The reason why there are missing surfaces on your marching cube result is because there is a very small space between two surfaces of the mesh. For example, the door has one surface inside, one surface outside of the cupboard, and there is only a small gap between them. So that some points in your grid may not fall between these surfaces, as a result, those points might be interpreted as points outside(negative sdf) the mesh.

So the quick solution would be to increase your grid resolution.
Hope this helps.
